common/src/main/java/com/wynntils/core/chat/transcoder/CodedString.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.chat.transcoder;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class CodedString {
+    private final List<CodedStringPart> parts;
+
+    private final List<ClickEvent> clickEvents = new LinkedList<>();
+    private final List<HoverEvent> hoverEvents = new LinkedList<>();
+
+    private Component componentCache;
+
+    CodedString(Component component) {
+        parts = new LinkedList<>();
+
+        CodedStringPart lastPart = null;
+
+        for (Component current : component.toFlatList()) {
+            final CodedStringPart finalLastPart = lastPart;
+
+            current.visit(
+                    (style, string) -> {
+                        parts.add(new CodedStringPart(string, style, this, finalLastPart));
+                        return Optional.empty();
+                    },
+                    Style.EMPTY);
+
+            lastPart = parts.get(parts.size() - 1);
+        }
+    }
+
+    // We don't want to expose the actual string to the outside world
+    // If you need to do an operation with this string, implement it as a method
+    private String getCoded() {
+        StringBuilder builder = new StringBuilder();
+
+        for (CodedStringPart part : parts) {
+            builder.append(part.getCoded());
+        }
+
+        return builder.toString();
+    }
+
+    public Component getComponent() {
+        if (componentCache != null) {
+            return componentCache;
+        }
+
+        MutableComponent component = Component.empty();
+
+        parts.forEach(part -> component.append(part.getComponent()));
+
+        componentCache = component;
+
+        return component;
+    }
+
+    public CodedStringPart getPartBefore(CodedStringPart codedStringPart) {
+        int index = parts.indexOf(codedStringPart);
```

Comment:
This depends on two CodedStringPart with the same content still not being "equal", like if you have a string: `§cHey §cHey §6Yo!`. Sounds a bit scary. I'm still trying to figure out where this is used, and if it can be solved differently.

Reply:
Why though? Every part is in the list, I am not sure how this is a problem.

Reply:
But the part index can change when we modify the whole string. I might as well save the component before this part directly, but that can still change.. And I think it would be a pain to keep track.